hilti/runtime/src/types/string.cc

```diff
@@ -132,35 +137,87 @@ std::tuple<std::string, std::string> string::split1(const std::string& s, const
 }
 
 Bytes string::encode(std::string s, unicode::Charset cs, unicode::DecodeErrorStrategy errors) {
+    if ( s.empty() )
+        return {std::move(s)};
+
     switch ( cs.value() ) {
         case unicode::Charset::UTF8: {
-            // Data supposedly is already in UTF-8, but let's validate it.
+            // HILTI `string` is always UTF-8, but we could be invoked with raw bags of bytes here as well, so validate.
             std::string t;
 
-            auto p = reinterpret_cast<const unsigned char*>(s.data());
-            auto e = p + s.size();
+            auto p = s.begin();
+            auto e = s.end();
 
             while ( p < e ) {
-                utf8proc_int32_t cp;
-                auto n = utf8proc_iterate(p, e - p, &cp);
+                try {
+                    auto cp = utf8::next(p, e);
+                    utf8::append(cp, t);
+                } catch ( const utf8::exception& ) {
+                    switch ( errors.value() ) {
+                        case unicode::DecodeErrorStrategy::IGNORE: break;
+                        case unicode::DecodeErrorStrategy::REPLACE: {
+                            utf8::append(unicode::REPLACEMENT_CHARACTER, t);
+                            break;
+                        }
+                        case unicode::DecodeErrorStrategy::STRICT:
+                            throw RuntimeError("illegal UTF8 sequence in string");
+                    }
+
+                    p = std::next(p);
+                }
+            }
+
+            return Bytes(std::move(t));
+        }
+
+        case unicode::Charset::UTF16BE: [[fallthrough]];
+        case unicode::Charset::UTF16LE: {
+            std::string t8;
 
-                if ( n < 0 ) {
+            auto p = s.begin();
+            auto e = s.end();
+
+            while ( p < e ) {
+                try {
+                    auto cp = utf8::next(p, e);
+                    utf8::append(cp, t8);
+                } catch ( const utf8::exception& ) {
                     switch ( errors.value() ) {
                         case unicode::DecodeErrorStrategy::IGNORE: break;
-                        case unicode::DecodeErrorStrategy::REPLACE: t += "\ufffd"; break;
+                        case unicode::DecodeErrorStrategy::REPLACE: {
+                            utf8::append(unicode::REPLACEMENT_CHARACTER, t8);
+                            break;
+                        }
                         case unicode::DecodeErrorStrategy::STRICT:
                             throw RuntimeError("illegal UTF8 sequence in string");
                     }
 
-                    p += 1;
-                    continue;
+                    p = std::next(p);
                 }
+            }
```

Comment:
This first part is just the UTF8 conversion/validation again, right? Better to factor that out than duplicating the code.

Reply:
Good idea, this can just call `string::encode` to UTF8.